File: MetaGeta/MetaGeta.DataStore/Model/File.vb
<Serializable()> _
Public Class MGFile
    Implements IEquatable(Of MGFile)

    Private m_ID As Long = -1
    Private ReadOnly m_DataStore As MGDataStore
    Private m_Tags As New MGTagCollection()

    Friend Sub New(ByVal datastore As MGDataStore)
        m_DataStore = datastore
    End Sub
    Friend Sub New(ByVal id As Long, ByVal datastore As MGDataStore)
        m_DataStore = datastore
        m_ID = id
    End Sub
    Public Sub New()
        Me.New(Nothing)
    End Sub

    Public Function GetTag(ByVal tagName As String) As String
        Return m_Tags.GetValue(tagName)
    End Function

    Public Sub SetTag(ByVal tagName As String, ByVal tagValue As String)
        m_Tags.SetValue(tagName, tagValue)
    End Sub

    Public Property ID() As Long
        Get
            Return m_ID
        End Get
        Set(ByVal value As Long)
            If m_ID <> value Then
                m_ID = value
            End If
        End Set
    End Property

    Public Overloads Function Equals(ByVal other As MGFile) As Boolean Implements System.IEquatable(Of MGFile).Equals
        Return other.ID = ID
    End Function

    Public ReadOnly Property FileName() As String
        Get
            Dim fn = GetTag(FileNameKey)
            If fn Is Nothing Then
                Throw New Exception("Can't find filename for file.")
            End If
            Return New Uri(fn).LocalPath
        End Get
    End Property

    Public ReadOnly Property Tags() As MGTagCollection
        Get
            Return m_Tags
        End Get
    End Property

#Region "Constants"
    Public Shared ReadOnly Property FileNameKey() As String
        Get
            Return "FileName"
        End Get
    End Property
    Public Shared ReadOnly Property TimeStampKey() As String
        Get
            Return "TimeStamp"
        End Get
    End Property
#End Region

End Class

Public Class MGFileIDComparer
    Implements IComparer(Of MGFile)

    Public Function Compare(ByVal x As MGFile, ByVal y As MGFile) As Integer Implements IComparer(Of MGFile).Compare
        Return x.ID.CompareTo(y.ID)
    End Function
End Class

Public Class MGFileEventArgs
    Inherits EventArgs

    Private ReadOnly m_File As MGFile

    Public Sub New(ByVal f As MGFile)
        m_File = f
    End Sub

    Public ReadOnly Property File() As MGFile
        Get
            Return m_File
        End Get
    End Property

End Class

File: MetaGeta/MetaGeta.DataStore/Model/File.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections.Generic;

#endregion

namespace MetaGeta.DataStore {
    [Serializable]
    public class MGFile : IEquatable<MGFile> {
        private readonly MGDataStore m_DataStore;

        private readonly MGTagCollection m_Tags;
        private long m_Id = -1;

        internal MGFile(MGDataStore datastore) {
            m_Tags = new MGTagCollection();
            m_DataStore = datastore;
        }

        internal MGFile(long id, MGDataStore datastore, MGTagCollection tags) {
            m_Tags = tags;
            m_DataStore = datastore;
            m_Id = id;
        }

        public MGFile() : this(null) {}

        public long ID {
            get { return m_Id; }
            set {
                if (m_Id != value)
                    m_Id = value;
            }
        }

        public string FileName {
            get {
                string fn = Tags.GetString(FileNameKey);
                if (fn == null)
                    throw new Exception("Can't find filename for file.");
                return new Uri(fn).LocalPath;
            }
        }

        public MGTagCollection Tags {
            get { return m_Tags; }
        }

        #region "Constants"

        public static string FileNameKey {
            get { return "FileName"; }
        }

        public static string TimeStampKey {
            get { return "TimeStamp"; }
        }

        #endregion

        #region IEquatable<MGFile> Members

        public bool Equals(MGFile other) {
            return other.ID == ID;
        }

        #endregion
    }
}

namespace MetaGeta.DataStore {
    public class MGFileIDComparer : IComparer<MGFile> {
        #region IComparer<MGFile> Members

        public int Compare(MGFile x, MGFile y) {
            return x.ID.CompareTo(y.ID);
        }

        #endregion
    }
}

namespace MetaGeta.DataStore {
    public class MGFileEventArgs : EventArgs {
        private readonly MGFile m_File;

        public MGFileEventArgs(MGFile f) {
            m_File = f;
        }

        public MGFile File {
            get { return m_File; }
        }
    }
}